core/arch/arm/kernel/thread.c

```diff
@@ -1411,26 +1412,69 @@ static struct mobj *thread_rpc_alloc_arg(size_t size)
 	return NULL;
 }
 
-static bool get_rpc_arg(uint32_t cmd, size_t num_params,
-			struct optee_msg_arg **arg_ret, uint64_t *carg_ret)
+static bool set_rmem(struct optee_msg_param *param,
+		     struct thread_param *tpm)
+{
+	param->attr = tpm->attr - THREAD_PARAM_ATTR_MEMREF_IN +
+		      OPTEE_MSG_ATTR_TYPE_RMEM_INPUT;
+	param->u.rmem.offs = tpm->u.memref.offs;
+	param->u.rmem.size = tpm->u.memref.size;
+	if (tpm->u.memref.mobj) {
+		param->u.rmem.shm_ref = mobj_get_cookie(tpm->u.memref.mobj);
+		if (param->u.rmem.shm_ref == MOBJ_INVALID_COOKIE)
+			return false;
+	} else {
+		param->u.rmem.shm_ref = 0;
+	}
+
+	return true;
+}
+
+static bool set_tmem(struct optee_msg_param *param,
+		     struct thread_param *tpm)
+{
+	paddr_t pa = 0;
+	uint64_t shm_ref = 0;
+	struct mobj *mobj = tpm->u.memref.mobj;
+
+	param->attr = tpm->attr - THREAD_PARAM_ATTR_MEMREF_IN +
+		      OPTEE_MSG_ATTR_TYPE_TMEM_INPUT;
+	if (mobj) {
+		shm_ref = mobj_get_cookie(mobj);
+		if (shm_ref == MOBJ_INVALID_COOKIE)
+			return false;
+		if (mobj_get_pa(mobj, tpm->u.memref.offs, 0, &pa))
+			return false;
+	}
+
+	param->u.tmem.size = tpm->u.memref.size;
+	param->u.tmem.buf_ptr = pa;
+	param->u.tmem.shm_ref = shm_ref;
+
+	return true;
+}
+
+static uint32_t get_rpc_arg(uint32_t cmd, size_t num_params,
```

Comment:
Should we replace return type `uint32_t` with `TEE_Result`?

Reply:
Sure, but I think it's better to do that after this PR.

Reply:
ok